src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaTypeName.cs

```diff
@@ -0,0 +1,151 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	// An internal Type that represents a parsed Java type name structure.
+	// A type name string is passed to Parse() method to create this instance
+	// (the only place you can create an instance of this type), then it is
+	// used to create JavaTypeReference instance.
+	//
+	// The structure does not depend on any context type information.
+	//
+	// It is constructed from a "dotted" name.
+	// Unlike JNI-based names, there is no way to determine that a dot is within
+	// package name, between package and class, or for nested class.
+	// Hence it is impossible to get a package and a local names without
+	// context type information. Such analyses should be done at JavaTypeReference.
+	//
+	// A generic type parameter, or a primitive type, is also represented by this too.
+	//
+	public class JavaTypeName
```

Comment:
I think I'm confused between the existence of `JavaTypeName` and `JavaTypeReference`.  The latter is "resolvable"; the former is not.  Is that a meaningful difference for API users?

Reply:
`JavaTypeName.cs` is scary, so it is c/p unchanged from `main` to ensure compatibility: https://github.com/xamarin/java.interop/blob/main/src/Xamarin.Android.Tools.ApiXmlAdjuster/JavaTypeName.cs.